File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.py

# def solution(participant, completion):
#     answer = set(participant) - set(completion)
#     return answer.pop()

# 동명이인이 있네;;
# 이러면 set 쓰면 안된다
def solution(participant, completion):
    part = {}
    total = 0
    for i in participant:
        part[hash(i)] = i
        total += hash(i)
    
    for j in completion:
        total -= hash(j)
    
    # total은 그냥 int 값
    return part[total]

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.*;
class Solution {
    public String solution(String[] participant, String[] completion) {
        Map<String, Integer> map = new HashMap<>();
        StringBuilder sb = new StringBuilder();
        
        for (int i = 0; i < participant.length; i++) {
            if (map.containsKey(participant[i])) {
                map.put(participant[i], map.get(participant[i]) + 1);
            } else {
                map.put(participant[i], 1);
            }
        }

        for (int i = 0; i < completion.length; i++) {
            if (map.containsKey(completion[i])) {
                map.put(completion[i], map.get(completion[i]) - 1);
                if (map.get(completion[i]) == 0) {
                    map.remove(completion[i]);
                }
            }
        }
        
        for (String s : map.keySet()) {
            sb.append(s);
        }
        
        return sb.toString();
    }
}